### ISSUE TITLE
Support fragmented Nginx configuration

### ISSUE THREAD

**Issue by harry-xm** [pappasam/nginx-language-server#11]
Common practice, as seen in default configuration on Ubuntu, is to split `server{}` blocks into separate files in the sites-enabled directory.

When I try to use nginx-language-server on these partial configuration files, crossplane could not correctly parse them because of errors like `"server" directive is not allowed here`.

Is it possible to detect such configuration and automatically add a surrounding block to fix crossplane parsing?

```
# /etc/nginx/sites-enabled/default
server {
	listen 80 default_server;
	listen [::]:80 default_server;
...
}

# nginx.conf
http {
...
	include /etc/nginx/conf.d/*.conf;
	include /etc/nginx/sites-enabled/*;
}
```

**Comment by pappasam** [comment]
Hmm, I'm not quite sure here. Will take some considerable digging. If you'd like to take a stab in the meantime, it looks like `crossplane.parse` takes a `combine` argument that's `False` by default. Setting it to `True` may get you on the right track.

**Comment by pappasam** [comment]
Actually, are you asking for completion within `/etc/nginx/sites-enabled/default`? In that case, I have no idea where to start there... if you have any ideas please let me know!

**Comment by pappasam** [comment]
Maybe we can do something with comments? We'd need some way to generally identify where an included script is intended to run. I think this would need to be communicated from the user to the language server somehow.

Alternatively, we could somehow identify whether there is a root `nginx.conf` and use that to determine the location of included files and figure out their context. Not sure what would be best here, will definitely take some design work.

**Comment by harry-xm** [comment]
> Actually, are you asking for completion within `/etc/nginx/sites-enabled/default`? In that case, I have no idea where to start there... if you have any ideas please let me know!

Yes. I did a quick hack, but it would break other configurations. I'm thinking of checking for `directive is not allowed here` error messages and redo parsing with such wrapping to see if those errors are gone.

```diff
diff --git a/nginx_language_server/parser/nginxconf.py b/nginx_language_server/parser/nginxconf.py
index 2480be3..fb11b36 100644
--- a/nginx_language_server/parser/nginxconf.py
+++ b/nginx_language_server/parser/nginxconf.py
@@ -42,8 +42,8 @@ def convert(code: str) -> ParsedSymbols:
     """Load and parse symbols into a container for future processing."""
     try:
         with NamedTemporaryFile(mode="w", delete=False) as nginx:
-            nginx.write(code)
-        parsed = crossplane.parse(nginx.name)["config"][0]["parsed"]
+            nginx.write("http{"+code+"}")
+        parsed = crossplane.parse(nginx.name, single=True)["config"][0]["parsed"]
         result: ParsedSymbols = {}
         _convert(result, parsed, [])
         return result
```

**Comment by harry-xm** [comment]
If crossplane supports parsing config within a context, the implementation would be much more elegant. I think it's worth asking.

**Comment by harry-xm** [comment]
We could use `check_ctx=False` to make `crossplane.parse()` work, and replace the default root context "main" with smart inference of the most probable context.

**Comment by pappasam** [comment]
Sounds like an interesting approach! Can't say I fully see a clear path forward, but I'll keep meditating and will happily review any PR's if you discover a viable path.

**Comment by ivucica** [comment]
I don't know what crossplane's output looks like, but would it make sense to, upon failure, retry by patching the input with `http{}` (possibly an option that can be switched off, or switched off unless folder name is `sites-available` or `sites-enabled`)?  That is: the same suggested by @harry-xm above, but perhaps only if the processing otherwise fails.

---

This would work for files embedded into `http{}` block. But, in addition to what's common on Debian, I also happen to have `modules-enabled`, `conf.d` and `streams-enabled` (to forward general TCP connections) -- these sometimes end up in different blocks / context.

So given that workspaces are a concept in LSP, for me, a good fix would be processing `${WORKSPACE_ROOT}/nginx.conf`, and possibly throwing away any result not in current file (thus still getting access to start col+row of the token and end col+row of the token, for highlighting and hovering and completion).  It's total guesswork that nginx.conf is intended, but it might work.

---

Or, as a different heuristic, detecting `/etc/nginx` is part of the path and assuming "ah, if this is the path, `/etc/nginx/nginx.conf is likely the real file to parse`".  It'd still depend on throwing away anything not in current file.

Does any of this make some sense?

---

Sidenote:  My environment happens to be `lsp-mode` in Emacs, and hovers and completions work fine in nginx.conf (or if I add `http{}` around the file manually). Thank you for your work, this is really a nice thing to see. I also happened to pull in `nginx-mode` for colorization prior to trying to use your language server; activating `nginx-mode` happens to be a good trigger for telling `lsp-mode` to use your server.
